util/tracecpu/collector.go

```diff
@@ -0,0 +1,391 @@
+	topSQLMap        map[string]*topSQLDataPoints
+	MaxStatementsNum int
+
+	// normalizedSQLMap is an map, whose keys are SQL digest strings and values are normalized SQL strings
+	normalizedSQLMap map[string]string
+
+	// normalizedPlanMap is a plan version of normalizedSQLMap
+	// this should only be set from the dedicated worker
+	normalizedPlanMap map[string]string
+
+	planRegisterChan chan *planRegisterJob
+
+	collectInterval  time.Duration
+	collectTimeout   time.Duration
+	agentGRPCAddress string
+
+	// current tidb-server instance ID
+	instanceID string
+
+	quit chan struct{}
+}
+
+// TopSQLCollectorConfig is the config for TopSQLCollector
+type TopSQLCollectorConfig struct {
+	// PlanBinaryDecoder is used to decode the plan binary before sending to agent
+	PlanBinaryDecoder planBinaryDecodeFunc
+	// MaxStatementsNum is the capacity of the TopSQL map, above which evition should happen
+	MaxStatementsNum int
+	// CollectInterval is the interval of sending TopSQL records to the agent
+	CollectInterval time.Duration
+	// CollectTimeout is the timeout of a single agent gRPC call
+	CollectTimeout time.Duration
+	// AgentGRPCAddress is the gRPC address of the agent server
+	AgentGRPCAddress string
+	// InstanceID is the ID of this tidb server
+	InstanceID string
+}
+
+func encodeCacheKey(sqlDigest, planDigest []byte) []byte {
+	var buffer bytes.Buffer
+	buffer.Write(sqlDigest)
+	buffer.Write(planDigest)
+	return buffer.Bytes()
+}
+
+const (
+	GrpcInitialWindowSize     = 1 << 30
+	GrpcInitialConnWindowSize = 1 << 30
+)
+
+func newAgentClient(addr string) (*grpc.ClientConn, tipb.TopSQLAgentClient, error) {
+	dialCtx, cancel := context.WithTimeout(context.TODO(), time.Second)
+	defer cancel()
+	conn, err := grpc.DialContext(
+		dialCtx,
+		addr,
+		grpc.WithInsecure(),
+		grpc.WithInitialWindowSize(GrpcInitialWindowSize),
+		grpc.WithInitialConnWindowSize(GrpcInitialConnWindowSize),
+		grpc.WithDefaultCallOptions(
+			grpc.MaxCallRecvMsgSize(math.MaxInt64),
+		),
+		grpc.WithConnectParams(grpc.ConnectParams{
+			Backoff: backoff.Config{
+				BaseDelay:  100 * time.Millisecond, // Default was 1s.
+				Multiplier: 1.6,                    // Default
+				Jitter:     0.2,                    // Default
+				MaxDelay:   3 * time.Second,        // Default was 120s.
+			},
+		}),
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	client := tipb.NewTopSQLAgentClient(conn)
+	return conn, client, nil
+}
+
+// NewTopSQLCollector creates a new TopSQL struct
+//
+// planBinaryDecoder is a decoding function which will be called asynchronously to decode the plan binary to string
+// MaxStatementsNum is the maximum SQL and plan number, which will restrict the memory usage of the internal LFU cache
+func NewTopSQLCollector(config *TopSQLCollectorConfig) *TopSQLCollector {
+	normalizedSQLMap := make(map[string]string)
+	normalizedPlanMap := make(map[string]string)
+	planRegisterChan := make(chan *planRegisterJob, 10)
```

Comment:
The buffer size `10` is too small. How about `1024*1024`?

Reply:
Will change to 1024 for now.

Reply:
ok, How about 1000, since the default, the max concurrency of execution SQL is 1000.